controllers/machine_controller_noderef.go

```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
 		return err
 	}
 
-	// Get the Node reference.
-	nodeRef, err := r.getNodeReference(remoteClient, providerID)
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
 	if err != nil {
 		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+				return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
+					"failed to find the Node owned by Machine %q in namespace %q", machine.Name, machine.Namespace)
+			}
+
+			logger.Error(err, "Failed to assign NodeRef")
+			r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
```

Comment:
We should probably keep logging and recording the event regardless of the error, like it was before. This behavior change might be worthwhile to investigate in v1alpha4